docs/decisions/0017-generalized-containers.rst

```diff
@@ -0,0 +1,81 @@
+
+3. Container Children and Relationships
+=======================================
+
+This section defines container children, their order, and relationships, covering flexible connections and support for draft and published states of their children.
+
+- Each container version holds a list of children that the author has defined for that version.
+- The author-defined list is used to show the content of a container version as the author specified it.
+- The author-defined list won't change for a specific container version even if its members change. E.g., a unit version UV1 with three components (CV1, CV2, CV3) will always have those three components in the author-defined list, even if one of the components is soft-deleted or a new version for the component is created.
+- The children of a container can be any type of publishable content. E.g., sections, subsections, units, components, and any other publishable thing. For more details on publishable content, see `PublishableEntity`_.
+- Children within a container are maintained in a specific order as an ordered list. E.g., components within a unit, or units within a subsection, are presented in a specific order.
+- Containers represent their content hierarchy through a structure, like Course > Section > Subsection > Unit > Component, which defines parent-child relationships at each level.
+- Containers can reference a specific version of their children or be set to point to their latest versions. For instance, component V1 might be used in a unit instead of its latest version. The latest version of a child can be referenced by setting its version to ``None`` which consists of the chosen standard for this representation.
+- A single child (publishable entity) can be shared by multiple containers, allowing for reuse of content across different containers. For instance, a component can be shared by multiple units.
+
+4. Container Versioning, Soft-Deletions and Structural Changes
+==============================================================
+
+This section defines the versioning rules for containers, explaining when a new version is created based on structural or metadata changes. It also clarifies how changes in child elements affect (or do not affect) container versions.
+
+- A new version of a container is created only when the container itself changes, such as modifying its title, ordering of children, or adding/removing children. Changes to a child do not cascade up, meaning modifications to a child do not trigger a new version of its parent. This includes the following cases:
+
+  - Soft-deleting a child that is referenced by a container does not create a new version of the container. For example, soft-deleting a component used by a unit will not generate a new version of the unit. Once the deletion is published, the child will be filtered out from the container as needed to ensure consistency.
```

Comment:
What is soft-deleting?

Reply:
Basically deleting. But we achieve it by setting the Draft to NULL, and when published this sets the published version to NULL, but all the historic versions are there and it can be un-deleted at any time by pointing the Draft back to any version (non-NULL).

Reply:
I added a key concepts section with those terms to (try) reducing ambiguity:  https://github.com/openedx/openedx-learning/pull/251/commits/20688f13ab7adaea0fd73d5d00f5b40c43d01a6c

Reply:
@kdmccormick We use "removal" and "deletion" for completely different things. Removal just means taking it off of some list (e.g. remove a unit from a container), but nothing is [soft] deleted.

> As a user, I remove a component from a unit within a library. This is a soft-deletion, right? Has the component also been removed from the library?

No, it's just removal, not deletion. The component continues to exist in the library and perhaps in other units.

> As a user, I remove a component from the library. 

There is no such concept as to "remove a component from the library". You may [soft] delete it however. Deleting a component is different than removing it from a unit.

Reply:
> There is no such concept as to "remove a component from the library"

But, from a user's perspective, it seems like there is such a concept, and it's we're calling it soft-deletion, no?

If not, then what does a user do in order to trigger the soft-deletion of something?

Reply:
@kdmccormick That's just called "deleting the component" (and in the code we call it "soft deletion" but that phrase is never used in the UI).

![Screenshot 2025-03-05 at 10 52 26 AM](https://github.com/user-attachments/assets/e14b48ad-3f67-4897-af99-52577be29786)

I just don't want to call that "removing" the component because it's doing a deletion, and in the unit context removal and deletion are very different things - the latter affects other stuff and the former only affects the unit you're removing from.

Reply:
OK, I understand now, ty.

Some terminology nits for the ADR, consider these all optional: the words "removal" and "(soft-)deletion" did not mean anything different to me until you explained the difference. And "soft-deletion" kept making me think that there'd be a such thing as "hard-deletion", which there isn't, unless that's what pruning is. I would just qualify it every time: "removal from the parent container", "deletion from the library". No need to specify "soft-" since IMO it's obvious that if we can undo a deletion prior to pruning, then the deletion does not actually involve immediately wiping the data from the database--any system that supports Undo would have to work this way.

Reply:
I agree that constantly using the term "soft deletion" is awkward. Those suggestions seem reasonable to me.